pkg/common/catalog/external.go

```diff
@@ -6,184 +6,138 @@ import (
 	"encoding/hex"
 	"errors"
 	"fmt"
-	"os"
 	"path/filepath"
 	"sync"
 
-	"github.com/hashicorp/go-hclog"
 	goplugin "github.com/hashicorp/go-plugin"
 	"github.com/sirupsen/logrus"
+	"github.com/spiffe/spire-plugin-sdk/private"
 	"github.com/spiffe/spire/pkg/common/log"
-	"github.com/spiffe/spire/pkg/common/telemetry"
 	"github.com/zeebo/errs"
 	"google.golang.org/grpc"
 )
 
-const (
-	// the ID used to dial host services
-	hostServicesID = 1
-)
-
-func PluginMain(plugin Plugin) {
-	logger := hclog.New(&hclog.LoggerOptions{
-		Level:      hclog.Trace,
-		Output:     os.Stderr,
-		JSONFormat: true,
-	})
-	goplugin.Serve(&goplugin.ServeConfig{
-		HandshakeConfig: goplugin.HandshakeConfig{
-			ProtocolVersion:  1,
-			MagicCookieKey:   plugin.Plugin.PluginType(),
-			MagicCookieValue: plugin.Plugin.PluginType(),
-		},
-		Plugins: map[string]goplugin.Plugin{
-			plugin.Name: &hcServerPlugin{
-				logger: logger,
-				plugin: plugin,
-			},
-		},
-		Logger:     logger,
-		GRPCServer: goplugin.DefaultGRPCServer,
-	})
-}
+type externalConfig struct {
+	// Name of the plugin
+	Name string
 
-type hcServerPlugin struct {
-	goplugin.NetRPCUnsupportedPlugin
-	logger hclog.Logger
-	plugin Plugin
-}
+	// Type is the plugin type (e.g. KeyManager)
+	Type string
 
-var _ goplugin.GRPCPlugin = (*hcServerPlugin)(nil)
+	// Path is the path on disk to the plugin.
+	Path string
 
-func (p *hcServerPlugin) GRPCServer(b *goplugin.GRPCBroker, s *grpc.Server) error {
-	initPluginServer(
-		s,
-		grpcBrokerDialer{b: b},
-		p.logger,
-		p.plugin.Plugin,
-		p.plugin.Services,
-	)
-	return nil
-}
+	// Args are the command line arguments to supply to the plugin
+	Args []string
 
-func (p *hcServerPlugin) GRPCClient(ctx context.Context, b *goplugin.GRPCBroker, c *grpc.ClientConn) (interface{}, error) {
-	return nil, errors.New("unimplemented")
-}
+	// Checksum is the hex-encoded SHA256 hash of the plugin binary.
+	Checksum string
 
-type grpcBrokerDialer struct {
-	b *goplugin.GRPCBroker
-}
+	// Log is the logger to be wired to the external plugin.
+	Log logrus.FieldLogger
 
-func (d grpcBrokerDialer) DialHost() (*grpc.ClientConn, error) {
-	return d.b.Dial(hostServicesID)
+	// HostServices are the host service servers provided to the plugin.
+	HostServices []HostServiceServer
 }
 
-type ExternalPlugin struct {
-	Log           logrus.FieldLogger
-	Name          string
-	Path          string
-	Checksum      string
-	Data          string
-	Plugin        PluginClient
-	KnownServices []ServiceClient
-	HostServices  []HostServiceServer
-}
+func loadExternal(ctx context.Context, config externalConfig) (*pluginImpl, error) {
+	// TODO: honor context cancellation... unfortunately go-plugin doesn't seem
+	// to give us a mechanism for this, so we'd have to spin up some goroutine
+	// to watch for cancelation and start killing clients and closing
```

Comment:
```suggestion
	// to watch for cancellation and start killing clients and closing
```

Reply:
Heh, both are correct but I'll yield to the one in more use :)

Reply:
And be internally consistent. Whoops. Had the other usage above in the same comment...